pkg/metrics3/client.go

```diff
@@ -0,0 +1,464 @@
+package metrics3
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	AccessKey           string = "api"
+	DataCollectionToken string = "data"
+)
+
+type AnodotToken struct {
+	Value string
+	Type  string
+}
+
+func NewAnoToken(token string, ttype string) (*AnodotToken, error) {
+	if len(strings.TrimSpace(token)) == 0 {
+		return nil, fmt.Errorf("token can't be blank")
+	}
+
+	switch ttype {
+	case AccessKey, DataCollectionToken:
+		return &AnodotToken{Value: token, Type: ttype}, nil
+	}
+
+	return nil, fmt.Errorf("token type can be DataCollectionToken or AccessKey")
+}
+
+type AnodotResponse interface {
+	HasErrors() bool
+	ErrorMessage() string
+	RawResponse() *http.Response
+}
+
+type Api30Response struct {
+	Error *struct {
+		Status        int    `json:"status"`
+		Name          string `json:"name"`
+		Message       string `json:"message"`
+		AndtErrorCode int    `json:"andtErrorCode"`
+		Path          string `json:"path"`
+	}
+	HttpResponse *http.Response `json:"-"`
+}
+
+func (r *Api30Response) HasErrors() bool {
+	return r.Error != nil
+}
+
+func (r *Api30Response) ErrorMessage() string {
+	return fmt.Sprintf("%+v\n", r.Error)
+}
+
+func (r *Api30Response) RawResponse() *http.Response {
+	return r.HttpResponse
+}
+
+type RefreshBearerResponse struct {
+	Bearer string
+	Api30Response
+}
+
+type Anodot30Client struct {
+	ServerURL   *url.URL
+	Token       *AnodotToken
+	client      *http.Client
+	BearerToken *struct {
+		timestemp time.Time
+		token     string
+	}
+}
+
+func NewAnodot30Client(anodotURL url.URL, token *AnodotToken, httpClient *http.Client) (*Anodot30Client, error) {
+	if token == nil {
+		return nil, fmt.Errorf("anodot token can't be nil")
+	}
+	submitter := Anodot30Client{Token: token, ServerURL: &anodotURL, client: httpClient, BearerToken: nil}
+	if httpClient == nil {
+		client := http.Client{Timeout: 30 * time.Second}
+
+		debugHTTP, _ := strconv.ParseBool(os.Getenv("ANODOT_HTTP_DEBUG_ENABLED"))
+		if debugHTTP {
+			client.Transport = &debugHTTPTransport{r: http.DefaultTransport}
+		}
+		submitter.client = &client
+	}
+
+	return &submitter, nil
+}
+
+func (c *Anodot30Client) GetBearerToken() (*string, error) {
+	// Token valid 24 hours, so if BearerToken field is null or token expired
+	// needs to refresh it, otherwise, returns existed token
+
+	if c.BearerToken == nil || time.Since(c.BearerToken.timestemp) > 24*time.Hour {
+		resp, err := c.refreshBearerToken()
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.HasErrors() {
+			return nil, fmt.Errorf("failed to refresh toke: %v", resp.ErrorMessage())
+		}
+
+		c.BearerToken = &struct {
+			timestemp time.Time
+			token     string
+		}{time.Now(), resp.Bearer}
```

Comment:
instead of `time.now` use `resp.refreshTime` ?

Reply:
Done